src/sentry/utils/iterators.py

```diff
@@ -1,5 +1,30 @@
 from __future__ import absolute_import
 
+import itertools
+
+
+def advance(n, iterator):
+    """Advances an iterator n places."""
+    next(itertools.islice(iterator, n, n), None)
+    return iterator
+
+
+def shingle(n, iterator):
```

Comment:
This implementation is pretty crafty with itertools building blocks and turning everything into an `izip`, but I thought due to the use of `itertools.islice` that it was `O(l^2)` when it shouldn't need to be more than `O(nl)`. I spent about 30 minutes convinced this was `O(l^2)` and trying to write an `O(nl)` implementation that didn't just yolocrutch on `list(iterator)` before I had refamiliarized myself with itertools enough to see that this actually *is* `O(nl)`. Might add a comment here explaining that this basically makes `n` sub-iterators, each offset by 0-n-1, and then zips those together n-wise to get each shingle, and it's `O(nl)`.